builtin/providers/aws/resource_aws_network_interface.go

```diff
@@ -0,0 +1,307 @@
+			Pending: []string{"true"},
+			Target:  "false",
+			Refresh: networkInterfaceAttachmentRefreshFunc(ec2conn, eniId),
+			Timeout: 10 * time.Minute,
+		}
+		if _, err := stateConf.WaitForState(); err != nil {
+		return fmt.Errorf(
+			"Error waiting for ENI (%s) to become dettached: %s", eniId, err)
+	}
+	}
+
+	return nil
+}
+
+func resourceAwsNetworkInterfaceUpdate(d *schema.ResourceData, meta interface{}) error {
+
+	d.Partial(true)
+
+	if d.HasChange("attachment") {
+		ec2conn := meta.(*AWSClient).awsEC2conn
+		oa, na := d.GetChange("attachment")			
+		
+		detach_err := resourceAwsNetworkInterfaceDetach(oa.(*schema.Set), meta, d.Id())
+		if detach_err != nil {
+			return detach_err
+		}
+
+		// if there is a new attachment, attach it
+		if na != nil && len(na.(*schema.Set).List()) > 0 {				
+			new_attachment := na.(*schema.Set).List()[0].(map[string]interface{})
+			attach_request := &ec2.AttachNetworkInterfaceRequest{
+				DeviceIndex:		aws.Integer(new_attachment["device_index"].(int)),
+				InstanceID:			aws.String(new_attachment["instance"].(string)),
+				NetworkInterfaceID:	aws.String(d.Id()),
+			}
+			_, attach_err := ec2conn.AttachNetworkInterface(attach_request)
+			if attach_err != nil {
+				return fmt.Errorf("Error attaching ENI: %s", attach_err)
+			}
+		}
+
+		d.SetPartial("attachment")
+	}
+
+	if d.HasChange("security_groups") {
+		request := &ec2.ModifyNetworkInterfaceAttributeRequest{
+			NetworkInterfaceID:		aws.String(d.Id()),
+			Groups:					expandStringList(d.Get("security_groups").(*schema.Set).List()),
+		}
+
+		ec2conn := meta.(*AWSClient).awsEC2conn
+		err := ec2conn.ModifyNetworkInterfaceAttribute(request)
+		if err != nil {
+			return fmt.Errorf("Failure updating ENI: %s", err)
+		}
+
+		d.SetPartial("security_groups")
+	}
+
+	d.Partial(false)
+
+	return resourceAwsNetworkInterfaceRead(d, meta)
+}
+
+func resourceAwsNetworkInterfaceDelete(d *schema.ResourceData, meta interface{}) error {
+	ec2conn := meta.(*AWSClient).awsEC2conn
+
+	log.Printf("[INFO] Deleting ENI: %s", d.Id())
+
+	detach_err := resourceAwsNetworkInterfaceDetach(d.Get("attachment").(*schema.Set), meta, d.Id())
+	if detach_err != nil {
+		return detach_err
+	}
+
+	deleteEniOpts := ec2.DeleteNetworkInterfaceRequest{
+		NetworkInterfaceID: aws.String(d.Id()),
+	}
+	if err := ec2conn.DeleteNetworkInterface(&deleteEniOpts); err != nil {
+		return fmt.Errorf("Error deleting ENI: %s", err)
+	}
+
+	return nil
+}
+
+func convertToJustAddresses(dtos []ec2.NetworkInterfacePrivateIPAddress) []string {
+	ips := make([]string, 0, len(dtos))
+	for _, v := range dtos {
+		ip := *v.PrivateIPAddress
+		ips = append(ips, ip)
+	}
+	return ips
+}
+
+func convertToGroupIds(dtos []ec2.GroupIdentifier) []string {
+	ids := make([]string, 0, len(dtos))
+	for _, v := range dtos {
+		group_id := *v.GroupID
+		ids = append(ids, group_id)
+	}
+	return ids
+}
+
+func convertToPrivateIPAddresses(ips []interface{}) []ec2.PrivateIPAddressSpecification {
```

Comment:
the convention we have for your conversion type methods is `flatten<things>` and `expand<things>`. You can see examples in [builtin/providers/aws/structure.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/structure.go) 

If possible, these should be unit tested as well, similar to [structure_test.go](https://github.com/hashicorp/terraform/blob/master/builtin/providers/aws/structure_test.go)